#  Web scraping with beautiful soup and cleaning in form of desired table

## First, we import necessary libraries

In [4]:

import pandas as pd

import seaborn as sns

import matplotlib.pyplot as plt

import numpy as np 

%matplotlib inline

In [5]:
from urllib.request import urlopen

In [6]:
from bs4 import BeautifulSoup

## the URL for the page containing the table is read 

In [7]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html = urlopen(url)

## pulling data out of HTML and XML files using beautiful soup library

In [8]:
soup = BeautifulSoup(html, 'lxml')
type (soup)

bs4.BeautifulSoup

In [9]:
title=soup.title
print (title)

<title>List of postal codes of Canada: M - Wikipedia</title>


## Reading each row of the table 

In [10]:
rows = soup.find_all('tr')
for row in rows:
    row_td = row.find_all('td')
print(row_td)
type(row_td)

[<td align="center" style="border: 1px solid #FF0000; background-color: #FFE0E0; font-size: 135%;" width="5%"><a href="/wiki/List_of_postal_codes_of_Canada:_A" title="List of postal codes of Canada: A">A</a>
</td>, <td align="center" style="border: 1px solid #FF4000; background-color: #FFE8E0; font-size: 135%;" width="5%"><a href="/wiki/List_of_postal_codes_of_Canada:_B" title="List of postal codes of Canada: B">B</a>
</td>, <td align="center" style="border: 1px solid #FF8000; background-color: #FFF0E0; font-size: 135%;" width="5%"><a href="/wiki/List_of_postal_codes_of_Canada:_C" title="List of postal codes of Canada: C">C</a>
</td>, <td align="center" style="border: 1px solid #FFC000; background-color: #FFF8E0; font-size: 135%;" width="5%"><a href="/wiki/List_of_postal_codes_of_Canada:_E" title="List of postal codes of Canada: E">E</a>
</td>, <td align="center" style="border: 1px solid #FFFF00; background-color: #FFFFE0; font-size: 135%;" width="5%"><a href="/wiki/List_of_postal_code

bs4.element.ResultSet

## extracting all the text in rows

In [11]:
str_cells = str(row_td)
cleantext = BeautifulSoup(str_cells, "lxml").get_text()
print(cleantext)

[A
, B
, C
, E
, G
, H
, J
, K
, L
, M
, N
, P
, R
, S
, T
, V
, X
, Y
]


In [12]:
import re

list_rows = []
for row in rows:
    cells = row.find_all('td')
    str_cells = str(cells)
    clean = re.compile('<.*?>')
    clean2 = (re.sub(clean, '',str_cells))
    list_rows.append(clean2)
print(clean2)
type(clean2)

[A
, B
, C
, E
, G
, H
, J
, K
, L
, M
, N
, P
, R
, S
, T
, V
, X
, Y
]


str

## transforming list of rows into dataframe form

In [13]:
df = pd.DataFrame(list_rows)
df.head(10)

0
0                                       []
1      [M1A, Not assigned, Not assigned\n]
2      [M2A, Not assigned, Not assigned\n]
3           [M3A, North York, Parkwoods\n]
4    [M4A, North York, Victoria Village\n]
5  [M5A, Downtown Toronto, Harbourfront\n]
6   [M5A, Downtown Toronto, Regent Park\n]
7    [M6A, North York, Lawrence Heights\n]
8      [M6A, North York, Lawrence Manor\n]
9      [M7A, Queen's Park, Not assigned\n]

## now the list has to be separated by commas

In [148]:
df1 = df[0].str.split(',',expand=True)
df1.head(10)

0                  1                     2     3     4     5     6   \
0    []               None                  None  None  None  None  None   
1  [M1A       Not assigned       Not assigned\n]  None  None  None  None   
2  [M2A       Not assigned       Not assigned\n]  None  None  None  None   
3  [M3A         North York          Parkwoods\n]  None  None  None  None   
4  [M4A         North York   Victoria Village\n]  None  None  None  None   
5  [M5A   Downtown Toronto       Harbourfront\n]  None  None  None  None   
6  [M5A   Downtown Toronto        Regent Park\n]  None  None  None  None   
7  [M6A         North York   Lawrence Heights\n]  None  None  None  None   
8  [M6A         North York     Lawrence Manor\n]  None  None  None  None   
9  [M7A       Queen's Park       Not assigned\n]  None  None  None  None   

     7     8     9   ...    21    22    23    24    25    26    27    28  \
0  None  None  None  ...  None  None  None  None  None  None  None  None   
1  None  None  None  ...  None  None  None  None  None  None  None  None   
2  None  None  None  ...  None  None  None  None  None  None  None  None   
3  None  None  None  ...  None  None  None  None  None  None  None  None   
4  None  None  None  ...  None  None  None  None  None  None  None  None   
5  None  None  None  ...  None  None  None  None  None  None  None  None   
6  None  None  None  ...  None  None  None  None  None  None  None  None   
7  None  None  None  ...  None  None  None  None  None  None  None  None   
8  None  None  None  ...  None  None  None  None  None  None  None  None   
9  None  None  None  ...  None  None  None  None  None  None  None  None   

     29    30  
0  None  None  
1  None  None  
2  None  None  
3  None  None  
4  None  None  
5  None  None  
6  None  None  
7  None  None  
8  None  None  
9  None  None  

[10 rows x 31 columns]

## Getting rid of unnecessary list characters such as [, ] and /n

In [149]:
df1[0] = df1[0].str.strip('[')
df1[2] = df1[2].str.strip('\n]')
df1.head(10)

0                  1                  2     3     4     5     6     7   \
0    ]               None               None  None  None  None  None  None   
1  M1A       Not assigned       Not assigned  None  None  None  None  None   
2  M2A       Not assigned       Not assigned  None  None  None  None  None   
3  M3A         North York          Parkwoods  None  None  None  None  None   
4  M4A         North York   Victoria Village  None  None  None  None  None   
5  M5A   Downtown Toronto       Harbourfront  None  None  None  None  None   
6  M5A   Downtown Toronto        Regent Park  None  None  None  None  None   
7  M6A         North York   Lawrence Heights  None  None  None  None  None   
8  M6A         North York     Lawrence Manor  None  None  None  None  None   
9  M7A       Queen's Park       Not assigned  None  None  None  None  None   

     8     9   ...    21    22    23    24    25    26    27    28    29    30  
0  None  None  ...  None  None  None  None  None  None  None  None  None  None  
1  None  None  ...  None  None  None  None  None  None  None  None  None  None  
2  None  None  ...  None  None  None  None  None  None  None  None  None  None  
3  None  None  ...  None  None  None  None  None  None  None  None  None  None  
4  None  None  ...  None  None  None  None  None  None  None  None  None  None  
5  None  None  ...  None  None  None  None  None  None  None  None  None  None  
6  None  None  ...  None  None  None  None  None  None  None  None  None  None  
7  None  None  ...  None  None  None  None  None  None  None  None  None  None  
8  None  None  ...  None  None  None  None  None  None  None  None  None  None  
9  None  None  ...  None  None  None  None  None  None  None  None  None  None  

[10 rows x 31 columns]

## filtering only first three columns

In [150]:
df1.drop(df1.columns.difference([0,1,2]), 1, inplace=True)

In [151]:
df1

0                  1  \
0                                                    ]               None   
1                                                  M1A       Not assigned   
2                                                  M2A       Not assigned   
3                                                  M3A         North York   
4                                                  M4A         North York   
5                                                  M5A   Downtown Toronto   
6                                                  M5A   Downtown Toronto   
7                                                  M6A         North York   
8                                                  M6A         North York   
9                                                  M7A       Queen's Park   
10                                                 M8A       Not assigned   
11                                                 M9A          Etobicoke   
12                                                 M1B        Scarborough   
13                                                 M1B        Scarborough   
14                                                 M2B       Not assigned   
15                                                 M3B         North York   
16                                                 M4B          East York   
17                                                 M4B          East York   
18                                                 M5B   Downtown Toronto   
19                                                 M5B   Downtown Toronto   
20                                                 M6B         North York   
21                                                 M7B       Not assigned   
22                                                 M8B       Not assigned   
23                                                 M9B          Etobicoke   
24                                                 M9B          Etobicoke   
25                                                 M9B          Etobicoke   
26                                                 M9B          Etobicoke   
27                                                 M9B          Etobicoke   
28                                                 M1C        Scarborough   
29                                                 M1C        Scarborough   
..                                                 ...                ...   
264                                                M5Y       Not assigned   
265                                                M6Y       Not assigned   
266                                                M7Y       East Toronto   
267                                                M8Y          Etobicoke   
268                                                M8Y          Etobicoke   
269                                                M8Y          Etobicoke   
270                                                M8Y          Etobicoke   
271                                                M8Y          Etobicoke   
272                                                M8Y          Etobicoke   
273                                                M8Y          Etobicoke   
274                                                M8Y          Etobicoke   
275                                                M9Y       Not assigned   
276                                                M1Z       Not assigned   
277                                                M2Z       Not assigned   
278                                                M3Z       Not assigned   
279                                                M4Z       Not assigned   
280                                                M5Z       Not assigned   
281                                                M6Z       Not assigned   
282                                                M7Z       Not assigned   
283                                                M8Z          Etobicoke   
284                                              

## dropping rows that are not data from table

In [152]:
df1.drop([0,289,290,291,292,293],inplace=True)

## deleting rows with 'Not assigned' borough

In [173]:
df2=df1[~df1[1].str.contains("Not")]

# when we look at length of neighborhoods ( i.e df[3]) the length of the strings show that there is an extra space :

In [193]:
df2[2].str.len()

3      10
4      17
5      13
6      12
7      17
8      15
9      13
11     17
12      6
13      8
15     16
16     17
17     14
18      8
19     16
20     10
23     11
24     10
25     13
26     17
27     16
28     15
29     11
30     11
32     16
33     16
34     17
35     15
36     19
39     18
       ..
236    12
237    16
240     9
241    32
244    10
245    12
246    10
247    12
250    12
251    15
252    21
253    17
256    13
257    16
258    15
263    21
266    50
267    11
268    17
269    25
270    10
271    15
272    19
273    22
274     9
283    25
284    10
285    19
286    22
287    15
Name: 2, Length: 211, dtype: int64

## Getting rid of that extra space

In [206]:
df2[2]=df2[2].str.strip()

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


## replacing the not assigned in neighborhood with its borough

In [217]:
df2.replace({2: {"Not assigned": "Queen's Park"}})

0                  1                                                  2
3    M3A         North York                                          Parkwoods
4    M4A         North York                                   Victoria Village
5    M5A   Downtown Toronto                                       Harbourfront
6    M5A   Downtown Toronto                                        Regent Park
7    M6A         North York                                   Lawrence Heights
8    M6A         North York                                     Lawrence Manor
9    M7A       Queen's Park                                       Queen's Park
11   M9A          Etobicoke                                   Islington Avenue
12   M1B        Scarborough                                              Rouge
13   M1B        Scarborough                                            Malvern
15   M3B         North York                                    Don Mills North
16   M4B          East York                                   Woodbine Gardens
17   M4B          East York                                      Parkview Hill
18   M5B   Downtown Toronto                                            Ryerson
19   M5B   Downtown Toronto                                    Garden District
20   M6B         North York                                          Glencairn
23   M9B          Etobicoke                                         Cloverdale
24   M9B          Etobicoke                                          Islington
25   M9B          Etobicoke                                       Martin Grove
26   M9B          Etobicoke                                   Princess Gardens
27   M9B          Etobicoke                                    West Deane Park
28   M1C        Scarborough                                     Highland Creek
29   M1C        Scarborough                                         Rouge Hill
30   M1C        Scarborough                                         Port Union
32   M3C         North York                                    Flemingdon Park
33   M3C         North York                                    Don Mills South
34   M4C          East York                                   Woodbine Heights
35   M5C   Downtown Toronto                                     St. James Town
36   M6C               York                                 Humewood-Cedarvale
39   M9C          Etobicoke                                  Bloordale Gardens
..   ...                ...                                                ...
236  M9V          Etobicoke                                        Thistletown
237  M1W        Scarborough                                    L'Amoreaux West
240  M4W   Downtown Toronto                                           Rosedale
241  M5W   Downtown Toronto                    Stn A PO Boxes 25 The Esplanade
244  M8W          Etobicoke                                          Alderwood
245  M8W          Etobicoke                                        Long Branch
246  M9W          Etobicoke                                          Northwest
247  M1X        Scarborough                                        Upper Rouge
250  M4X   Downtown Toronto                                        Cabbagetown
251  M4X   Downtown Toronto                                     St. James Town
252  M5X   Downtown Toronto                               First Canadian Place
253  M5X   Downtown Toronto                                   Underground city
256  M8X          Etobicoke                                       The Kingsway
257  M8X          Etobicoke                                    Montgomery Road
258  M8X          Etobicoke                                     Old Mill North
263  M4Y   Downtown Toronto                               Church and Wellesley
266  M7Y       East Toronto  Business Reply Mail Processing Centre 969 Eastern
267  M8Y          Etobicoke                                         Humber Bay
268  M8Y          Etobicoke                              

## Now we group the data by postalcode and borough and combine all rows with same postalcode and borough

In [285]:
df3 = df2.groupby([0, 1],as_index=False).agg(lambda x: x.tolist())

In [286]:
df3

0                  1                                                  2
0    M1B        Scarborough                                   [Rouge, Malvern]
1    M1C        Scarborough           [Highland Creek, Rouge Hill, Port Union]
2    M1E        Scarborough                [Guildwood, Morningside, West Hill]
3    M1G        Scarborough                                           [Woburn]
4    M1H        Scarborough                                        [Cedarbrae]
5    M1J        Scarborough                              [Scarborough Village]
6    M1K        Scarborough      [East Birchmount Park, Ionview, Kennedy Park]
7    M1L        Scarborough                  [Clairlea, Golden Mile, Oakridge]
8    M1M        Scarborough  [Cliffcrest, Cliffside, Scarborough Village West]
9    M1N        Scarborough                      [Birch Cliff, Cliffside West]
10   M1P        Scarborough  [Dorset Park, Scarborough Town Centre, Wexford...
11   M1R        Scarborough                                [Maryvale, Wexford]
12   M1S        Scarborough                                        [Agincourt]
13   M1T        Scarborough          [Clarks Corners, Sullivan, Tam O'Shanter]
14   M1V        Scarborough  [Agincourt North, L'Amoreaux East, Milliken, S...
15   M1W        Scarborough                                  [L'Amoreaux West]
16   M1X        Scarborough                                      [Upper Rouge]
17   M2H         North York                                [Hillcrest Village]
18   M2J         North York                     [Fairview, Henry Farm, Oriole]
19   M2K         North York                                  [Bayview Village]
20   M2L         North York                         [Silver Hills, York Mills]
21   M2M         North York                          [Newtonbrook, Willowdale]
22   M2N         North York                                 [Willowdale South]
23   M2P         North York                                  [York Mills West]
24   M2R         North York                                  [Willowdale West]
25   M3A         North York                                        [Parkwoods]
26   M3B         North York                                  [Don Mills North]
27   M3C         North York                 [Flemingdon Park, Don Mills South]
28   M3H         North York  [Bathurst Manor, Downsview North, Wilson Heights]
29   M3J         North York                  [Northwood Park, York University]
..   ...                ...                                                ...
73   M6C               York                               [Humewood-Cedarvale]
74   M6E               York                              [Caledonia-Fairbanks]
75   M6G   Downtown Toronto                                         [Christie]
76   M6H       West Toronto                     [Dovercourt Village, Dufferin]
77   M6J       West Toronto                         [Little Portugal, Trinity]
78   M6K       West Toronto     [Brockton, Exhibition Place, Parkdale Village]
79   M6L         North York               [Downsview, North Park, Upwood Park]
80   M6M               York   [Del Ray, Keelesdale, Mount Dennis, Silverthorn]
81   M6N               York                    [The Junction North, Runnymede]
82   M6P       West Toronto                    [High Park, The Junction South]
83   M6R       West Toronto                           [Parkdale, Roncesvalles]
84   M6S       West Toronto                               [Runnymede, Swansea]
85   M7A       Queen's Park                                     [Not assigned]
86   M7R        Mississauga            [Canada Post Gateway Processing Centre]
87   M7Y       East Toronto  [Business Reply Mail Processing Centre 969 Eas...
88   M8V          Etobicoke     [Humber Bay Shores, Mimico South, New Toronto]
89   M8W          Etobicoke                           [Alderwood, Long Branch]
90   M8X          Etobicoke    [The Kingsway, Montgomery Road, Old Mill North]
91   M8Y          Etobicoke  [Humber Bay, King's Mill Par

## I probably should have renamed columns earlier, but it was  a lot of work dealing with typing names. So, now the columns are renamed to proper names:

In [288]:
df3.columns=['PostalCode','Borough','Neighborhood']

In [289]:
df3

PostalCode            Borough  \
0          M1B        Scarborough   
1          M1C        Scarborough   
2          M1E        Scarborough   
3          M1G        Scarborough   
4          M1H        Scarborough   
5          M1J        Scarborough   
6          M1K        Scarborough   
7          M1L        Scarborough   
8          M1M        Scarborough   
9          M1N        Scarborough   
10         M1P        Scarborough   
11         M1R        Scarborough   
12         M1S        Scarborough   
13         M1T        Scarborough   
14         M1V        Scarborough   
15         M1W        Scarborough   
16         M1X        Scarborough   
17         M2H         North York   
18         M2J         North York   
19         M2K         North York   
20         M2L         North York   
21         M2M         North York   
22         M2N         North York   
23         M2P         North York   
24         M2R         North York   
25         M3A         North York   
26         M3B         North York   
27         M3C         North York   
28         M3H         North York   
29         M3J         North York   
..         ...                ...   
73         M6C               York   
74         M6E               York   
75         M6G   Downtown Toronto   
76         M6H       West Toronto   
77         M6J       West Toronto   
78         M6K       West Toronto   
79         M6L         North York   
80         M6M               York   
81         M6N               York   
82         M6P       West Toronto   
83         M6R       West Toronto   
84         M6S       West Toronto   
85         M7A       Queen's Park   
86         M7R        Mississauga   
87         M7Y       East Toronto   
88         M8V          Etobicoke   
89         M8W          Etobicoke   
90         M8X          Etobicoke   
91         M8Y          Etobicoke   
92         M8Z          Etobicoke   
93         M9A          Etobicoke   
94         M9B          Etobicoke   
95         M9C          Etobicoke   
96         M9L         North York   
97         M9M         North York   
98         M9N               York   
99         M9P          Etobicoke   
100        M9R          Etobicoke   
101        M9V          Etobicoke   
102        M9W          Etobicoke   

                                          Neighborhood  
0                                     [Rouge, Malvern]  
1             [Highland Creek, Rouge Hill, Port Union]  
2                  [Guildwood, Morningside, West Hill]  
3                                             [Woburn]  
4                                          [Cedarbrae]  
5                                [Scarborough Village]  
6        [East Birchmount Park, Ionview, Kennedy Park]  
7                    [Clairlea, Golden Mile, Oakridge]  
8    [Cliffcrest, Cliffside, Scarborough Village West]  
9                        [Birch Cliff, Cliffside West]  
10   [Dorset Park, Scarborough Town Centre, Wexford...  
11                                 [Maryvale, Wexford]  
12                                         [Agincourt]  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  
15                                   [L'Amoreaux West]  
16                                       [Upper Rouge]  
17                                 [Hillcrest Village]  
18                      [Fairview, Henry Farm, Oriole]  
19                                   [Bayview Village]  
20                          [Silver Hills, York Mills]  
21                           [Newtonbrook, Willowdale]  
22                                  [Willowdale South]  
23                                   [York Mills West]  
24                                   [Willowdale West]  
25                                         [Parkwoods]  
26                                   [Don Mills North]  
27                  [Flemingdon Park, Don Mills South]  
28   [Bathurst Manor, Downsview North, Wilson Heights]  

## our table is ready! here is the shape of it:

In [291]:
df3.shape

(103, 3)

## Let's read the CSV file containting geospatial coordinates 

In [292]:
dfgeo=pd.read_csv('Geospatial_Coordinates.csv')

In [294]:
dfgeo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## The cell below loops through the scraped table and when it finds them in geospatial dataframe, it appends the corresponding latitude and longitude to latlistand longlist.

In [313]:
latlist=[]
longlist=[]
for index, row in df3.iterrows():
        if (df3.loc[index, "PostalCode"]==dfgeo.loc[index,"Postal Code"]):
            latlist.append(dfgeo.loc[index,"Latitude"])
            longlist.append(dfgeo.loc[index,"Longitude"])

## To make sure the loop was correct, let's check if out latlist has the same number of elements as our scraped table (103)

In [314]:
len(latlist)

103

## Perfect! Let's add latitude and longitude to our table 

In [316]:
df3['Latitude']=latlist

In [317]:
df3['Longitude']=longlist

In [318]:
df3

PostalCode            Borough  \
0          M1B        Scarborough   
1          M1C        Scarborough   
2          M1E        Scarborough   
3          M1G        Scarborough   
4          M1H        Scarborough   
5          M1J        Scarborough   
6          M1K        Scarborough   
7          M1L        Scarborough   
8          M1M        Scarborough   
9          M1N        Scarborough   
10         M1P        Scarborough   
11         M1R        Scarborough   
12         M1S        Scarborough   
13         M1T        Scarborough   
14         M1V        Scarborough   
15         M1W        Scarborough   
16         M1X        Scarborough   
17         M2H         North York   
18         M2J         North York   
19         M2K         North York   
20         M2L         North York   
21         M2M         North York   
22         M2N         North York   
23         M2P         North York   
24         M2R         North York   
25         M3A         North York   
26         M3B         North York   
27         M3C         North York   
28         M3H         North York   
29         M3J         North York   
..         ...                ...   
73         M6C               York   
74         M6E               York   
75         M6G   Downtown Toronto   
76         M6H       West Toronto   
77         M6J       West Toronto   
78         M6K       West Toronto   
79         M6L         North York   
80         M6M               York   
81         M6N               York   
82         M6P       West Toronto   
83         M6R       West Toronto   
84         M6S       West Toronto   
85         M7A       Queen's Park   
86         M7R        Mississauga   
87         M7Y       East Toronto   
88         M8V          Etobicoke   
89         M8W          Etobicoke   
90         M8X          Etobicoke   
91         M8Y          Etobicoke   
92         M8Z          Etobicoke   
93         M9A          Etobicoke   
94         M9B          Etobicoke   
95         M9C          Etobicoke   
96         M9L         North York   
97         M9M         North York   
98         M9N               York   
99         M9P          Etobicoke   
100        M9R          Etobicoke   
101        M9V          Etobicoke   
102        M9W          Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                     [Rouge, Malvern]  43.806686 -79.194353  
1             [Highland Creek, Rouge Hill, Port Union]  43.784535 -79.160497  
2                  [Guildwood, Morningside, West Hill]  43.763573 -79.188711  
3                                             [Woburn]  43.770992 -79.216917  
4                                          [Cedarbrae]  43.773136 -79.239476  
5                                [Scarborough Village]  43.744734 -79.239476  
6        [East Birchmount Park, Ionview, Kennedy Park]  43.727929 -79.262029  
7                    [Clairlea, Golden Mile, Oakridge]  43.711112 -79.284577  
8    [Cliffcrest, Cliffside, Scarborough Village West]  43.716316 -79.239476  
9                        [Birch Cliff, Cliffside West]  43.692657 -79.264848  
10   [Dorset Park, Scarborough Town Centre, Wexford...  43.757410 -79.273304  
11                                 [Maryvale, Wexford]  43.750072 -79.295849  
12                                         [Agincourt]  43.794200 -79.262029  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  43.781638 -79.304302  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  43.815252 -79.284577  
15                                   [L'Amoreaux West]  43.799525 -79.318389  
16                                       [Upper Rouge]  43.836125 -79.205636  
17                                 [Hillcrest Village]  43.803762 -79.363452  
18                      [Fairview, Henry Farm, Oriole]  43.778517 -79.346556  
19                                   [Bayview Village]  43.786947 -79.385975  
20                          [Silver Hills, York Mi

## Now let's explore the neighborhoods containing the word "Toronto"


In [382]:
df4= df3[df3['Borough'].str.contains("Toronto")]

In [383]:
df4

PostalCode            Borough  \
37        M4E       East Toronto   
41        M4K       East Toronto   
42        M4L       East Toronto   
43        M4M       East Toronto   
44        M4N    Central Toronto   
45        M4P    Central Toronto   
46        M4R    Central Toronto   
47        M4S    Central Toronto   
48        M4T    Central Toronto   
49        M4V    Central Toronto   
50        M4W   Downtown Toronto   
51        M4X   Downtown Toronto   
52        M4Y   Downtown Toronto   
53        M5A   Downtown Toronto   
54        M5B   Downtown Toronto   
55        M5C   Downtown Toronto   
56        M5E   Downtown Toronto   
57        M5G   Downtown Toronto   
58        M5H   Downtown Toronto   
59        M5J   Downtown Toronto   
60        M5K   Downtown Toronto   
61        M5L   Downtown Toronto   
63        M5N    Central Toronto   
64        M5P    Central Toronto   
65        M5R    Central Toronto   
66        M5S   Downtown Toronto   
67        M5T   Downtown Toronto   
68        M5V   Downtown Toronto   
69        M5W   Downtown Toronto   
70        M5X   Downtown Toronto   
75        M6G   Downtown Toronto   
76        M6H       West Toronto   
77        M6J       West Toronto   
78        M6K       West Toronto   
82        M6P       West Toronto   
83        M6R       West Toronto   
84        M6S       West Toronto   
87        M7Y       East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                      [The Beaches]  43.676357 -79.293031  
41                     [The Danforth West, Riverdale]  43.679557 -79.352188  
42                   [The Beaches West, India Bazaar]  43.668999 -79.315572  
43                                  [Studio District]  43.659526 -79.340923  
44                                    [Lawrence Park]  43.728020 -79.388790  
45                                 [Davisville North]  43.712751 -79.390197  
46                               [North Toronto West]  43.715383 -79.405678  
47                                       [Davisville]  43.704324 -79.388790  
48                      [Moore Park, Summerhill East]  43.689574 -79.383160  
49  [Deer Park, Forest Hill SE, Rathnelly, South H...  43.686412 -79.400049  
50                                         [Rosedale]  43.679563 -79.377529  
51                      [Cabbagetown, St. James Town]  43.667967 -79.367675  
52                             [Church and Wellesley]  43.665860 -79.383160  
53                        [Harbourfront, Regent Park]  43.654260 -79.360636  
54                         [Ryerson, Garden District]  43.657162 -79.378937  
55                                   [St. James Town]  43.651494 -79.375418  
56                                      [Berczy Park]  43.644771 -79.373306  
57                               [Central Bay Street]  43.657952 -79.387383  
58                         [Adelaide, King, Richmond]  43.650571 -79.384568  
59  [Harbourfront East, Toronto Islands, Union Sta...  43.640816 -79.381752  
60         [Design Exchange, Toronto Dominion Centre]  43.647177 -79.381576  
61                   [Commerce Court, Victoria Hotel]  43.648198 -79.379817  
63                                         [Roselawn]  43.711695 -79.416936  
64              [Forest Hill North, Forest Hill West]  43.696948 -79.411307  
65              [The Annex, North Midtown, Yorkville]  43.672710 -79.405678  
66                   [Harbord, University of Toronto]  43.662696 -79.400049  
67        [Chinatown, Grange Park, Kensington Market]  43.653206 -79.400049  
68  [CN Tower, Bathurst Quay, Island airport, Harb...  43.628947 -79.394420  
69                  [Stn A PO Boxes 25 The Esplanade]  43.646435 -79.374846  
70           [First Canadian Place, Underground city]  43.648429 -79.382280  
75                                         [Christie]  43.669542 -79.422564  
76                     [Dovercourt Village, Dufferin]  43.669005 -79.442259  
77                      

In [384]:
from geopy.geocoders import Nominatim
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 43.653963, -79.387207.


In [390]:
import folium
# create map of New York using latitude and longitude values
map_Toronto= folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, postalcode in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['PostalCode']):
    label = '{}, {}'.format(borough, postalcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

In [391]:
CLIENT_ID = 'ASRI2J3DNFY1NYGNCINKNUFVWMU2JAHIQ52BGIZZ2ZRI1ZAG' # your Foursquare ID
CLIENT_SECRET = 'JCTR5JLVL31YS4PB0BM2BGTVIKBN2DJW5HR4LOORMI2ZEIGB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ASRI2J3DNFY1NYGNCINKNUFVWMU2JAHIQ52BGIZZ2ZRI1ZAG
CLIENT_SECRET:JCTR5JLVL31YS4PB0BM2BGTVIKBN2DJW5HR4LOORMI2ZEIGB
